pkg/util/service_failure_watcher.go

```diff
@@ -0,0 +1,37 @@
+package util
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+// ServiceFailureWatcher waits for service failures, and passed them to the channel.
+type ServiceFailureWatcher struct {
+	ch chan error
+}
+
+func NewServiceFailureWatcher() *ServiceFailureWatcher {
+	return &ServiceFailureWatcher{ch: make(chan error)}
+}
+
+// Returns channel for this watcher. If watcher is nil, returns nil channel.
+// Errors returned on the channel include failure case and service description.
+func (w *ServiceFailureWatcher) Chan() <-chan error {
+	if w == nil {
+		return nil
+	}
+	return w.ch
+}
+
+func (w *ServiceFailureWatcher) WatchService(service services.Service) {
+	service.AddListener(services.NewListener(nil, nil, nil, nil, func(from services.State, failure error) {
+		w.ch <- errors.Wrapf(failure, "service %v failed", service)
```

Comment:
What if, due to misuse or bug, no gorouting is reading on the chan? Given each listener has its own goroutine, this shouldn't be a problem. May you just confirm it, please?

Reply:
That is correct. Goroutine handling the listener events will be blocked forever though. We can use some small buffer to help with that [although it can still occur, if more errors happen], or ignore it, as we expect to server to shutdown soon after anyway.